File: Chess_Robot/src/stockfish/Stockfish.java
package stockfish;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;

public class Stockfish {

	private Process engineProcess;
	private BufferedReader processReader;
	private OutputStreamWriter processWriter;

	private static final String PATH = "./engine/stockfish.exe";

	public boolean startEngine() {
		try {
			engineProcess = Runtime.getRuntime().exec(PATH);
			processReader = new BufferedReader(new InputStreamReader(
					engineProcess.getInputStream()));
			processWriter = new OutputStreamWriter(
					engineProcess.getOutputStream());
		} catch (Exception e) {
			return false;
		}
		return true;
	}

	public void stopEngine() {
		try {
			sendCommand("quit");
			processReader.close();
			processWriter.close();
		} catch (Exception e) {
		}
	}

	public void sendCommand(String command) {
		try {
			processWriter.write(command + "\n");
			processWriter.flush();
		} catch (Exception e) {
		}
	}

	public String getOutput(int waitTime) {
		StringBuffer buffer = new StringBuffer();

		try {
			Thread.sleep(waitTime);
			sendCommand("isready");
			while (true) {
				String output = processReader.readLine();
				if (output.equals("readyok"))
					break;
				else
					buffer.append(output + "\n");
			}
		} catch (Exception e) {
		}
		return buffer.toString();
	}

	public String getBestMove(String fen, int waitTime) {
		sendCommand("go movetime " + waitTime);
		return getOutput(waitTime + 20).split("bestmove ")[1].substring(0, 4);
	}

	public String getFen() {
		sendCommand("d");
		return getOutput(0).split("Fen: ")[1].split("\n")[0];
	}

	public String moveAndGetFen(String fen, String move) {
		sendCommand("position fen " + fen + " moves " + move);
		return getFen();
	}
}

File: server/chessboard/chessRobot/stockfish.py
import subprocess
import time
from pathlib import Path

curr_path = Path(__file__).parent

class Stockfish:
    def __init__(self):
        self.engineProcess = None
        self.processReader = None
        self.processWriter = None
        self.PATH = [curr_path / './engine/stockfish.exe']

    def startEngine(self):
        try:
            self.engineProcess = subprocess.Popen(self.PATH, stdin=subprocess.PIPE, stdout=subprocess.PIPE, universal_newlines=True)
            self.processReader = self.engineProcess.stdout
            self.processWriter = self.engineProcess.stdin
        except Exception as e:
            print(e)
            return False
        return True

    def sendCommand(self, command):
        try:
            self.processWriter.write(command + "\n")
            self.processWriter.flush()
        except Exception as e:
            print(e)

    def stopEngine(self):
        try:
            self.sendCommand("quit")
            self.processReader.close()
            self.processWriter.close()
        except Exception as e:
            print(e)

    def getOutput(self, waitTime, bestmove):
        buffer = ""
        check1 = not bestmove
        check2 = False

        try:
            time.sleep(waitTime / 1000)
            self.sendCommand("isready")
            while True:
                output = self.processReader.readline().rstrip()

                if "bestmove" in output:
                    check1 = True
                if "readyok" in output:
                    check2 = True

                buffer += output + " \n"
                if check1 and check2:
                    break
        except Exception as e:
            print(e)

        return buffer

    def getBestMove(self, waitTime):
        self.sendCommand("go movetime " + str(waitTime))
        output = self.getOutput(waitTime, True)
        return output.split("bestmove ")[1].split(" ")[0]

    def getFen(self):
        self.sendCommand("d")
        output = self.getOutput(0, False)
        return output.split("Fen: ")[1].split(" \n")[0]

    def moveAndGetFen(self, fen, move):
        self.sendCommand("position fen " + fen + " moves " + move)
        return self.getFen()
